homework/zad9_opencv/line_match2.cpp

```diff
@@ -0,0 +1,161 @@
+#include <opencv2/imgproc.hpp>
+#include <opencv2/highgui.hpp>
+
+#include <iostream>
+#include <mutex>
+#include <future>
+#include <thread>
+
+void show_wait_destroy(const char* winname, cv::Mat img) {
+    cv::imshow(winname, img);
+    cv::moveWindow(winname, 500, 0);
+    cv::waitKey(0);
+    cv::destroyWindow(winname);
+}
+
+cv::Mat createMatLineAlpha( const cv::Size& size, int iAlpha )
+{
+    cv::Mat mat(size, CV_8U);
+    cv::Point origin( size.width / 2, size.height / 2 );
+    double dRad = ( iAlpha * M_PI ) / 180.0;
+
+    if ( iAlpha != 90 )
+    {
+        for( int iCol = 0; iCol < size.width; ++iCol )
+        {
+            double dValue = std::tan(dRad) * ( iCol - origin.x );
+            int iPos = origin.y - std::lround(dValue);
+
+            if ( iPos >= 0 && iPos < size.height )
+            {
+                mat.at<unsigned char>( iPos , iCol) = 1;
+            }
+        }
+    }
+    else
+    {
+        for ( int iRow = 0; iRow < size.height; ++iRow )
+        {
+            mat.at<unsigned char>(iRow, origin.x) = 1;
+        }
+    }
+
+    return mat;
+}
+
+auto lineMatchFunction = [](const cv::Mat& bw_image, const cv::Size& size, int alpha)
+{
+    // Create structure element for extracting lines through morphology operations
+    cv::Mat imgStructure = createMatLineAlpha(size, alpha);
+    cv::Mat imgLocal = bw_image.clone();
+
+    // Apply morphology operations
+    cv::erode(imgLocal, imgLocal, imgStructure, cv::Point(-1, -1));
+    cv::dilate(imgLocal, imgLocal, imgStructure, cv::Point(-1, -1));
+
+    return imgLocal.clone();
+};
+
+int main( int argc, char** argv )
+{
+    cv::CommandLineParser parser(argc, argv, "{@input | notes.png | input image}");
+    cv::Mat src = cv::imread( cv::samples::findFile( parser.get<cv::String>("@input") ), cv::IMREAD_COLOR);
+    if (src.empty())
+    {
+        std::cout << "Could not open or find the image!\n" << std::endl;
+        std::cout << "Usage: " << argv[0] << " <Input image>" << std::endl;
+        return -1;
+    }
+
+    // Show source image
+    cv::imshow("src", src);
+
+    // Transform source image to gray if it is not already
+    cv::Mat gray;
+    if (src.channels() == 3)
+    {
+        cvtColor(src, gray, cv::COLOR_BGR2GRAY);
+    }
+    else
+    {
+        gray = src;
+    }
+
+    // Show gray image
+    show_wait_destroy("gray", gray);
+
+    // Apply adaptiveThreshold at the bitwise_not of gray, notice the ~ symbol
+    cv::Mat bw;
+    cv::adaptiveThreshold(~gray, bw, 255, cv::ADAPTIVE_THRESH_MEAN_C, cv::THRESH_BINARY, 15, -2);
+    
+    // Show binary image
+    show_wait_destroy("binary", bw);
+
+    // Create the images that will use to extract the angle lines
+    cv::Mat matAngle0  = bw.clone();
+    cv::Mat matAngle30 = bw.clone();
+    cv::Mat matAngle45 = bw.clone();
+    cv::Mat matAngle60 = bw.clone();
+    cv::Mat matAngle90 = bw.clone();
+    cv::Mat matAngle135 = bw.clone();
+    cv::Mat matAngle150 = bw.clone();
```

Comment:
to wszystko się wydaje niepotrzebne, bo i tak wewnątrz `lineMatchFunction` robisz clone.

Reply:
ok